keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1069,187 @@ the remaining requests in that queue start getting faster service.  In
+However, given we don't want to block introducing the `width` concept on
+design and implementation of borrowing, until this is done we have two
+main options:
+- cap the `width` at the concurrency units assigned to the priority level
+- reject requests for which we won't be able to allocate enough concurrency
+  units
+
+To avoid breaking some users, we will proceed with the first option.  That
+said, to accommodate for the inaccuracy here we will introduce a concept
+of `additional latency` for a request.  This basically means that after the
+request finishes in a real world, we still don't mark it as finished in
+the virtual world for `additional latency`.
+Adjusting virtual time of a queue to do that is trivial.  The other thing
+to tweak is to ensure that the concurrency units will not get available
+for other requests for that time (because currently all actions are
+triggerred by starting or finishing some request).  We will maintain that
+possibility by wrapping the handler into another one that will be sleeping
+for `additional latence` after the request is processed.
+
+Note that given the estimation for duration of processing the requests is
+automatically corrected (both up and down), there is no need to change that
+in the initial version.
+
+#### Dispatching the request
+
+The hardest part of adding support for LIST requests is dispatching the
+requests.  Now in order to start processing a request, it has to accumulate
+`<width>` units of concurrency.
+
+The important requirement to recast now is fairness.  As soon a single
+request can consume more units of concurrency, the fairness is
+no longer about the number of requests from a given queue, but rather
+about number of consumed concurrency units.  This justifes the above
+definition of adjusting the cost of the request to now be equal to
+`<width> x <processing latency>` (instead of just `<processing latency>`).
+
+At the same time, we want to maximally utilize the available capacity.
+In other words, we want to minimize the time when some concurrency unit
+is not used, but there are requests at a given PL that could use it.
+
+In order to achieve the above goals, we are introducing the following
+modification to the current dispatching algorith:
+- as soon as we dispatch the first token (concurrency unit) to a queue
+  with LIST request at its front, all further tokens are also assigned
+  to this queue until the LIST request can be started.
+  Such approach allows us to not block too many tokens and avoid deadlocks
+  if we would be dispatching tokens to multiple LIST requests without having
+  enough of them for a given priority level.
+- however, to ensure fairness (especially over longer period of times)
+  the virtual time of the queue is now advanced by the new cost of the
+  request, i.e. by `<width> x <processing latency>` (as opposed to simply
+  `<processing latency>`).
+  This can be thought of as a negative budget that we accumulate (due to
```

Comment:
I think what I suggested earlier is a better first policy, since it will be a much smaller change, let me restate it:

Today, we already have a fairness ordering over the queues, and notion of how many seats there are. I think the minimal change is to track currently-executing-width rather than currently-occupied-seats. Don't change the order over the queues. Accept a request from the next queue as long as currently-executing-width < available-width.

This will let us (temporarily) exceed the width budget, which IMO is fine. It will potentially block multiple narrow requests while a single very wide request is executing. I think this is probably not optimal from a queuing theory perspective. Instead of theorizing about the best way to fix it, I think we should get this feature into a state where we can just directly measure improvements. (Although, I personally expect it will actually be non-trivial to beat the policy I'm suggesting here.)

Reply:
> This will let us (temporarily) exceed the width budget, which IMO is fine.

My main point is that I think it's not. Because with such policy we can potentially admit more requests than we'are able to reasonably process. And the main goal of this feature is overload protection. 


I realized that maybe what I'm writing is not clear. I'm not really suggesting to change the order in which we chose the next queue. All I'm saying is that:
- if we should choose a request from the queue and there are less than <width> available seat, we don't dispatch *any* request, until we will have <width> available seats when we dispatch this request

Reply:
Yes what you wrote in the comment is much clearer, I would move that into the text :)

I don't have a strong opinion re: waiting for N seats to be empty vs going over and then blocking the next few requests until at least one seat frees up. I think both would be better than what we do today. I have advocated for the latter mostly because you then don't need to cap the width.

Reply:
Sure, I agree with that. The priority ordering might make a good addition to the KEP's goals (maybe not in this PR :) ).

Reply:
Opened https://github.com/kubernetes/enhancements/pull/2682.